Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1375 @@
+
+```hs
+instance Applicative (Parser i e) where
+  pure a = Parser $ \input offset -> Right (offset, a, input)
+
+  Parser f <*> Parser p = Parser $ \input offset ->
+    case f input offset of
+      Left err -> Left err
+      Right (offset', f', rest) ->
+        case p rest offset' of
+          Left err -> Left err
+          Right (offset'', output, rest') -> Right (offset'', f' output, rest')
+```
+
+Finally, in `parse`, use `0` as the initial offset.
+
+Keep in mind it's also possible to operate with line and column when you know you have a function operating in characters.
+
+```hs
+>>> parse (string "hey") "hi"
+Left [Error {erOffset = 1, erError = Expected 'e' 'i'}]
+```
+
+The reader may also choose to instead refactor the parser type to use a custom-defined `State` type that contains the input stream and offset. Many improvements are possible, but we choose to keep it simple for educational purposes.
+  <hr>
+</details>
+
+### What's next?
+
+In this section, we've learned how to make a simple parsing combinator library. There are still various improvements that can be done. To name a few of them, we can implement more instances, such as `Semigroup` and `Monoid`, `IsString`, `MonadFail`, `MonadPlus`, among others. Error messages can be improved by better keeping track of location (such as line and column), supporting more types such as `Text` and `ByteString`, pretty-printing error messages, etc.
+
+We will conclude this part here, and if you are interested, proceed to the next part where Megaparsec, an industrial-strength parser combinator library, will be introduced with a more practical application than our simple examples in this section.
+
+## Megaparsec tutorial
+
+<!--
+* explain why we chose megaparsec
+* parse S-exps
+* make a small format for data parsing, no programming constructs
+  * equivalent to JSON: objects, lists, null, numbers, booleans, strings
+-->
+
+We will now take a look at a practical application with parser combinators: parsing S-expressions. Hopefully, such examples will give you a solid foundation for creating parsers for your grammars.
+
+Megaparsec is the go-to parser for industrial projects, so we will also use it in our tutorial. If you wish to see how it compares to other parser combinator libraries in Haskell, you can read its [`README.md`](https://github.com/mrkkrp/megaparsec#comparison-with-other-solutions).
+
+Keep in mind you should choose the parser library that better suits your needs. If you plan to parse machine-written data, for example, then Attoparsec may be a better choice. Its interface is similar to Megaparsec's, so the knowledge you will learn here will also be useful.
+
+Without any further ado, we will now present how to use Megaparsec to parse some [`S-expressions`](https://en.wikipedia.org/wiki/S-expressions).
```

Comment:
I would link in the first paragraph, since people will probably Google when seeing the first occurance.

Reply:
I linked in the first paragraph and removed the code-font.